### ISSUE TITLE
CRLF-Bug on Windows

### ISSUE THREAD

**Issue by MattDiMu** [stylelint/stylelint#1081]
As reported in the corresponding atom-plugin (linter-stylelint) in issue https://github.com/AtomLinter/linter-stylelint/issues/155 there is an error with CRLF-files on Windows, where stylelint seems to think, that the line ending has already been reached (even it hasn't). You will find more detailed info in the link above.

edit: I've further investigated the issue and it seems to occur only if the following conditions are both true
- Line ending is CRLF
- certain linting errors like function-max-empty-lines exist (maybe because this is an error, which should be shown at the end of line?)

When using stylelint on the command line, the issue is not so critical, as "only" the numbers are wrong. When used in conjunction with https://github.com/AtomLinter/linter-stylelint it will, however, break the plugin and make it quite unusable.

**Comment by evilebottnawi** [comment]
@MattDiMu Can you formalize issue as template and show example code with issue?

**Comment by MattDiMu** [comment]
Sorry for not being specific enough. You may reproduce the problem by following these steps (using default rules in .stylelintrc)

- Create _test.scss, use LF as line endings.
```scss
.thislinehas27characters {


}
```

- run stylelint (syntax-parameter is not necessary, result is the same)
```cmd
stylelint _test.scss --syntax=scss

_test.scss
1:26    Unexpected empty block (block-no-empty)
1:27    Unexpected empty line (max-empty-lines)
```

- Convert to CRLF line endings and run the same command again
```cmd
stylelint _test.scss --syntax=scss

_test.scss
1:26    Unexpected empty block (block-no-empty)
1:28    Unexpected empty line (max-empty-lines)
```

As you can see, it shows the non-existent **28**th character. This is probably the problem for  AtomLinter/linter-stylelint#155

**Comment by jeddy3** [comment]
@MattDiMu Thanks for the detailed write-up. This could be a tricky one to track down as it might be specific to a particular parser. Can you help us by trying your tests again but without the the `--syntax=scss` option? That will at least help us determine if the problem is in the scss parser or within one of the stylelint rules.

**Comment by MattDiMu** [comment]
@jeddy3 As already written above not using the --syntax option produces the same result. So it's probably not an scss parser issue.

**Comment by davidtheclark** [comment]
I suspect that what's happening is the rule is noting the location of `\n`, whereas I guess it should note the location of `\r` when it encounters `\r\n`.

**Comment by davidtheclark** [comment]
@MattDiMu if you can possibly come up with a list of which rules have this problem, we can try to make sure to address them all. We have two above.

**Comment by davidtheclark** [comment]
@MattDiMu Can you please try the branch over here https://github.com/stylelint/stylelint/pull/1094 and verify that it fixes the problem for `max-empty-lines`? If it does, we can implement the same little fix for other rules with the same problem.

**Comment by MattDiMu** [comment]
@davidtheclark - Sorry for bothering you, but how may i use the "crlf-line-ending-fix"-branch? It's not on the npm-registery and when simply cloning the repo, i don't get the compiled stylelint (dist-version).

**Comment by davidtheclark** [comment]
You could clone the repo into your `node_modules` directory; checkout the
right branch; then move into it and `npm install`, which will get the
dependencies and build it. I think that works.

**Comment by MattDiMu** [comment]
Unfortunatly, even after multiple attempts (upgrading/downgrading npm and 2 different OS), I've failed to install the stylelint devDependencies correctly. This might be because of my settings (disabled post-install-scripts), but I won't change that due to security reasons. Sorry, but I won't be able to help you test this fix.

**Comment by davidtheclark** [comment]
Ok. Well I'm pretty sure that PR fixes the problem for max-empty-lines, specifically. We just have to figure out which other rules to target.

**Comment by jeddy3** [comment]
> Well I'm pretty sure that PR fixes the problem for max-empty-lines, specifically

I just tested `master` locally, which has the fix for `max-empty-lines` in but not for `function-max-length-lines`. The column numbers for the former stayed the same when I switched from LF to CRLF, where as the latter increased by one. So, it looks to have worked.

> We just have to figure out which other rules to target.

Any idea how best to go about that?

**Comment by davidtheclark** [comment]
We need to find the rules that search for `\n` and then attach a warning to that particular character.

The tests for max-empty-lines actually showed that we were expecting a different column number for CRLF than for LF code snippets. But we didn't know it would be an issue that would break editors at that point.

Could also search tests for "CRLF", since we often included that in descriptions. Or search tests for `\r\n`.

**Comment by davidtheclark** [comment]
I did a quick run-through of the rules and don't see any others that I can tell at a glance would cause this issue. Not exactly sure how to seal the deal here ... I guess release the fixes we did and see if that means people have no trouble with the standard config?

**Comment by jeddy3** [comment]
> I guess release the fixes we did and see if that means people have no trouble with the standard config?

Sounds like a plan to me.

**Comment by davidtheclark** [comment]
If you want to release that'd be swell, because I can't get to it until later tonight.

**Comment by davidtheclark** [comment]
Or I'll just get to it later.

**Comment by jeddy3** [comment]
We'll need to wait for the rebase and merge of https://github.com/stylelint/stylelint/pull/1095, right?

If that happens while I'm still awake, I'll gladly release :)

**Comment by jeddy3** [comment]
Released in [6.0.3](https://github.com/stylelint/stylelint/releases/tag/6.0.3)
